// baojs/dist/router/router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tree_1 = require("./tree");
const httpMethods = [
    "GET",
    "HEAD",
    "POST",
    "PUT",
    "DELETE",
    "CONNECT",
    "OPTIONS",
    "TRACE",
    "PATCH",
];
const NOT_FOUND = { handler: null, params: {} };
class BaoRouter {
    constructor(opts = {}) {
        if (!(this instanceof BaoRouter)) {
            return new BaoRouter(opts);
        }
        this.trees = {};
        this.opts = opts;
    }
    any(path, handler) {
        for (const method of httpMethods) {
            this.on(method, path, handler);
        }
    }
    on(method, path, handler) {
        if (path[0] !== "/") {
            throw new Error("path must begin with '/' in path");
        }
        if (!this.trees[method]) {
            this.trees[method] = new tree_1.default();
        }
        this.trees[method].addRoute(path, handler);
        return this;
    }
    find(method, path) {
        const tree = this.trees[method];
        if (tree) {
            return tree.search(path);
        }
        return NOT_FOUND;
    }
}
exports.default = BaoRouter;
